### DISCUSSION TITLE
support self hosted?

### DISCUSSION THREAD

**Discussion by gd4Ark** [ankitpokhrel/jira-cli#271] [Ideas] [answered]
Didn't see about self hosted, Whether there is scheduled support?  thx!

**Comment by ankitpokhrel** [comment, marked as answer]
Hi @gd4Ark self hosted jira is already supported, see [on-premise installation](https://github.com/ankitpokhrel/jira-cli#on-premise-installation). I am not sure what you mean by scheduled support, could you clarify a bit please?

**Reply by gd4Ark** [reply]
It's my neglect, sorry

**Reply by gd4Ark** [reply]
I'm using jira version v8.13.x, but [Access Tokens](https://confluence.atlassian.com/enterprise/using-personal-access-tokens-1026032365.html) is only supported in 8.14 and later, so I had to miss this perfect tool

**Reply by ankitpokhrel** [reply]
Hi @gd4Ark, may be the doc was not clear enough. You don't need PAT. For self-hosted installation, you can use your regular jira password that you use for logging-in to the UI.

```
export JIRA_API_TOKEN=<your-regular-jira-password>
```

**Reply by gd4Ark** [reply]
Yes, I tried it. but I got 401. I'm sure the password is correct.

I run `jira init --debug` got:
```bash
? Installation type: Local
? Link to Jira server: https://jira.xxx.com
? Login username: xxx
⠋ Verifying login details...

REQUEST DETAILS
------------------------------------------------------------

GET /rest/api/2/myself HTTP/1.1
Host: jira.xxx.com
Authorization: Basic xxxxxxxxxxxxxxxxxxxxxxxxxxx=



RESPONSE DETAILS
------------------------------------------------------------

HTTP/1.1 401
Transfer-Encoding: chunked
Content-Security-Policy: frame-ancestors 'self'
Content-Type: text/html;charset=UTF-8
Date: Thu, 17 Feb 2022 15:27:01 GMT
Referrer-Policy: strict-origin-when-cross-origin
Set-Cookie: JSESSIONID=xxxxxxxxxxxxxxxxxxxxxxxxxx; Path=/; HttpOnly
Strict-Transport-Security: max-age=10368000
Www-Authenticate: OAuth realm="https%3A%2F%2Fjira.xxx.com"
X-Arequestid: 1407x4002230x1
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
X-Seraph-Loginreason: AUTHENTICATED_FAILED
X-Xss-Protection: 1; mode=block


✗ Received unexpected response '401 ' from jira. Please try again.
```

I see website same api different request header is work:
```bash
curl --request GET \curl 'https://jira.xxx.com/rest/api/2/myself' \
  -H 'authority: jira.xxx.com' \
  -H 'pragma: no-cache' \
  -H 'cache-control: no-cache' \
  -H 'sec-ch-ua: "Microsoft Edge";v="93", " Not;A Brand";v="99", "Chromium";v="93"' \
  -H 'accept: application/json, application/vnd-ms-excel' \
  -H 'x-atlassian-mau-ignore: true' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'content-type: application/json' \
  -H 'user-agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/93.0.4577.63 Safari/537.36 Edg/93.0.961.47' \
  -H 'sec-ch-ua-platform: "macOS"' \
  -H 'sec-fetch-site: same-origin' \
  -H 'sec-fetch-mode: cors' \
  -H 'sec-fetch-dest: empty' \
  -H 'referer: https://jira.xxx.com/secure/RapidBoard.jspa?rapidView=46' \
  -H 'accept-language: zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6' \
  -H 'cookie: xxx' \
  -H 'dnt: 1' \
  -H 'sec-gpc: 1' \
  --compressed
  --url 'https://jira.xxx.com/rest/api/2/myself' \
  --user "xxx:my_password"
```

I try remove cookie got 401:
```bash
curl --request GET \                                                                                                                 4s
  --url 'https://jira.xxx.com/rest/api/2/myself' \
  --user "xxx:my_password"
```

So, I think not support my jira version: v8.13.10#813010-sha1:156ab00

Maybe the administrator banned this operation?

**Reply by ankitpokhrel** [reply]
@gd4Ark I just checked in v8.13.1 and everything seems to be working fine. So, either this is something related to your installation or the exported variable is not available in your shell.

For the initial part, I cannot help you much. You can check with your administrator.

For the later part, make sure that `JIRA_API_TOKEN` is available in your shell with correct value before you execute `jira init`. If you are using a fish shell, you can follow [fishshell docs](https://fishshell.com/docs/2.3/tutorial.html#tut_exports).

**Reply by gd4Ark** [reply]
Thx, I sure my shell variable  is available
```bash
> env | grep "JIRA"                                                                                                                                            21s
JIRA_API_TOKEN=xxx
```

I think login my company selfhost  jira must set some cookie.

**Comment by MattPD** [comment]
Is there a way to use jira-cli with on-premise installation that also uses its own TLS certificates?
Simply attempting  `jira-cli init` currently fails with the following message:
`✗ Unable to generate configuration: x509: certificate signed by unknown authority`

Using ` jira init --debug`:
```
? Installation type: Local
? Link to Jira server: https://MYJIRAINSTANCE
? Login username: MYUSERNAME
⠹ Verifying login details... panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x77a3d7]

goroutine 1 [running]:
net/http/httputil.DumpResponse(0x0, 0x0)
        /usr/local/Cellar/go/1.17.5/libexec/src/net/http/httputil/dump.go:311 +0x37
github.com/ankitpokhrel/jira-cli/pkg/jira.dump(0xc0003543c0, 0xc000a92400)
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/pkg/jira/client.go:234 +0x45
github.com/ankitpokhrel/jira-cli/pkg/jira.(*Client).request.func1()
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/pkg/jira/client.go:217 +0x2e
github.com/ankitpokhrel/jira-cli/pkg/jira.(*Client).request(0xc00030a550, {0xe75fd0, 0xc0000301a0}, {0xc6c6fe, 0x3}, {0xc00040e580, 0x37}, {0x0, 0x0, 0x0}, ...)
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/pkg/jira/client.go:229 +0x55f
github.com/ankitpokhrel/jira-cli/pkg/jira.(*Client).GetV2(...)
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/pkg/jira/client.go:150
github.com/ankitpokhrel/jira-cli/pkg/jira.(*Client).Me(0xc00030a550)
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/pkg/jira/me.go:17 +0xcc
github.com/ankitpokhrel/jira-cli/internal/config.(*JiraCLIConfig).verifyLoginDetails(0xc000b3fcf8, {0xc0003ee000, 0xc000a6c8d0}, {0xc0005829e0, 0xc000a1a2c0})
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/internal/config/generator.go:227 +0x185
github.com/ankitpokhrel/jira-cli/internal/config.(*JiraCLIConfig).configureServerAndLoginDetails(0xc000b3fcf8)
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/internal/config/generator.go:215 +0x3db
github.com/ankitpokhrel/jira-cli/internal/config.(*JiraCLIConfig).Generate(0x0)
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/internal/config/generator.go:81 +0x54
github.com/ankitpokhrel/jira-cli/internal/cmd/init.initialize(0xc000a54f00, {0xc6fcfa, 0x1, 0x1})
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/internal/cmd/init/init.go:29 +0x7d
github.com/spf13/cobra.(*Command).execute(0xc000a54f00, {0xc000a6c760, 0x1, 0x1})
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/vendor/github.com/spf13/cobra/command.go:860 +0x5f8
github.com/spf13/cobra.(*Command).ExecuteC(0xc000a54c80)
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/vendor/github.com/spf13/cobra/command.go:974 +0x3bc
main.main()
        /Users/ankitpokhrel/go/src/github.com/ankitpokhrel/jira-cli/cmd/jira/main.go:12 +0x1e
```

FWIW, setting `JIRA_API_TOKEN` to my API token doesn't have any effect.

It seems to be a similar issue to https://github.com/andygrunwald/go-jira/issues/52 although it's not clear whether any of the workarounds mentioned in the thread apply to jira-cli.

**Reply by ankitpokhrel** [reply]
Hi @MattPD, can you please install the tool using the command below and see if it works? I've applied the solution mentioned in `go-jira` thread. If that works I will add a CLI/config option to control `InsecureSkipVerify`

```sh
go install github.com/ankitpokhrel/jira-cli/cmd/jira@patch-tls-issue
```

**Reply by MattPD** [reply]
Thank you! I can confirm this works (together with `JIRA_API_TOKEN` environmental variable using the value I've created via the Jira Web interface--clicking "Create API token" in my profile settings).

**Reply by pr-tkarner** [reply]
Hey @ankitpokhrel ,

the patch-tls-issue version works perfectly - is the config option already available in the latest version?

**Reply by ankitpokhrel** [reply]
@pr-tkarner this is merged in `main` branch and will be available in the next release. If you are using homebrew, you can install main branch with `brew install jira-cli --HEAD`. You can also compile the tool locally with `make install`.

You need to pass `--insecure` option during initialization step to make it work.

```sh
jira init --insecure
```

**Reply by pushkargr** [reply]
Hey @ankitpokhrel is this change available in the version 1.0.0?  I tried the most recent docker image and still seeing the issue.  `brew install jira-cli --HEAD` works fine though but does not satisfy my use case. 
```bash
 ❯ docker run -it --rm ghcr.io/ankitpokhrel/jira-cli:latest
~ # export JIRA_API_TOKEN=<redacted>
~ # jira init --insecure
You are using --insecure option. In this mode, the client will NOT verify
server's certificate chain and host name in requests to the jira server.

? Installation type: Local
? Link to Jira server: <redacted>
? Login username: <redacted>

✗ Unable to generate configuration: x509: certificate signed by unknown authority
```

**Reply by ankitpokhrel** [reply]
Hi @pushkargr, it looks like an issue with config generator. Does it work for you if you manually add this line in your config yaml file?

```
insecure: true
```

**Reply by pushkargr** [reply]
This worked. Thanks. I think I'm unblocked for now. Thanks @ankitpokhrel and great work.
